File: Rook.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace chess
{
    public class Rook : Piece
    {
        public Rook(Square position, string color)
        {
            Position = position;
            Color = color;
            Content = Color == "White" ? "\u2656" : "\u265C";
            FirstMove = true;
        }
        public Square Position { get; set; }
        public string Content { get; set; }
        public string Color { get; set; }
        public bool FirstMove { get; set; }

        public bool IsAvailableMove(Square target)
        {
            if (FirstMove)
            {
                return (target.Rank == Position.Rank && target.File != Position.File)
                    || (target.File == Position.File && target.Rank != Position.Rank);
            }
            else
            {
                return (target.Rank == Position.Rank && target.File != Position.File)
                 || (target.File == Position.File && target.Rank != Position.Rank);
            }
        }
        public bool CanCapture(Square target)
        {
            return IsAvailableMove(target);
        }
        public void Move(Square target)
        {
            Position.Rank = target.Rank;
            Position.File = target.File;
        }
    }
}

File: chess-cross-platform/chess-qt/rook.cpp
#include <string>
#include "rook.h"
#include "square.h"

Rook::Rook(Square* position, std::string color) {
  this->Position = position;
  this->Color = color;
  this->Content = Color == "White" ? "\u2656" : "\u265C";
  this->FirstMove = true;
}

Rook::~Rook(){}

bool Rook::IsAvailableMove(Square* target) {
  if (this->FirstMove)
    {
      return (target->Rank == Position->Rank && target->File != Position->File)
        || (target->File == Position->File && target->Rank != Position->Rank);
    }
  else
    {
      return (target->Rank == Position->Rank && target->File != Position->File)
        || (target->File == Position->File && target->Rank != Position->Rank);
    }
}

bool Rook::CanCapture(Square* target) {
  return IsAvailableMove(target);
}

void Rook::Move(Square* target) {
  Position->Rank = target->Rank;
  Position->File = target->File;
}
